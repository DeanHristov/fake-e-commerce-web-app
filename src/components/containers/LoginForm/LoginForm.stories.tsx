import type { Meta, StoryObj } from '@storybook/react';
import LoginForm, { ILoginFormProps } from './LoginForm';

const meta: Meta<ILoginFormProps> = {
  title: 'Component/LoginForm',
  component: LoginForm,
  parameters: {
    layout: 'centered',
  },

  argTypes: {},
};

type Story = StoryObj<ILoginFormProps>;

export const Primary: Story = {
  args: {},
};

export default meta;
