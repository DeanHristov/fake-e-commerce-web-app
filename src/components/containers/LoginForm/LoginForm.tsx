'use client';

import { APIUtils } from '@/utils/APIUtils';
import { useFormik } from 'formik';
import Link from 'next/link';
import { useRouter } from 'next/navigation';
import { FC, useState } from 'react';
import { toast } from 'react-toastify';
import { object, string } from 'yup';
import Button from '../../ui/Button';
import FormField from '../../ui/FormField';
import Widget from '../Widget';

export interface ILoginFormProps {}

export enum LOGIN_FORM_FIELDS {
  EMAIL = 'email',
  PASSWORD = 'password',
}

// TODO Adjust the form with an AWS Cognito workflow!
// TODO Make the form more user-friendly with a nice feedback on un-success sign in
// TODO Fix the issue in console: content.js:1 Uncaught (in promise) Error: Something went wrong. Please check back shortly.
const LoginForm: FC<ILoginFormProps> = ({}) => {
  const router = useRouter();
  const [spinning, setSpinning] = useState<boolean>(false);
  const formik = useFormik({
    validateOnChange: false,
    initialValues: {
      [LOGIN_FORM_FIELDS.EMAIL]: 'visitor@my-site.com',
      [LOGIN_FORM_FIELDS.PASSWORD]: '13579',
    },

    validationSchema: object({
      [LOGIN_FORM_FIELDS.EMAIL]: string().email().required(),
      [LOGIN_FORM_FIELDS.PASSWORD]: string().min(3).max(33).required(),
    }),

    onSubmit: async (values, { resetForm }) => {
      setSpinning(true);

      const { statusCode, error } = await APIUtils.fetch(
        `${process.env.BASE_URL}/sign-in/`,
        {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
          },
          body: JSON.stringify(values),
        },
      );

      setSpinning(false);
      if (statusCode === 400 || error) {
        resetForm();
        return toast.error(error, {
          position: toast.POSITION.TOP_RIGHT,
          autoClose: false,
        });
      }

      router.push('/');
      router.refresh();
    },
  });

  return (
    <Widget title="Login">
      <form onSubmit={formik.handleSubmit}>
        <FormField
          className="py-2 w-96"
          label="Enter an Email"
          id={LOGIN_FORM_FIELDS.EMAIL}
          name={LOGIN_FORM_FIELDS.EMAIL}
          value={formik.values[LOGIN_FORM_FIELDS.EMAIL]}
          onChange={formik.handleChange}
        />

        <FormField
          className="py-2 w-96"
          label="Enter a password"
          type="password"
          id={LOGIN_FORM_FIELDS.PASSWORD}
          name={LOGIN_FORM_FIELDS.PASSWORD}
          value={formik.values[LOGIN_FORM_FIELDS.PASSWORD]}
          onChange={formik.handleChange}
        />

        <Button
          type="submit"
          spinning={spinning}
          className="py-2 w-96 rounded-xl"
          title={'Login'}
        />
      </form>

      <Link
        href="/reset-password"
        className="text-blue-400 text-sm mt-4 text-center hover:underline"
      >
        Reset password
      </Link>
    </Widget>
  );
};

export default LoginForm;
