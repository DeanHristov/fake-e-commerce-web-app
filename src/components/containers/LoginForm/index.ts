import LoginForm, { ILoginFormProps } from './LoginForm';

export type { ILoginFormProps };

export default LoginForm;
