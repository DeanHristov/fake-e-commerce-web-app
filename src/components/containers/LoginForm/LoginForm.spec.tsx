import { render } from '@testing-library/react';
import LoginForm from './LoginForm';

describe('Container/Component <LoginForm {...} />', () => {
  it('Should render the badge with a counter', () => {
    const { container } = render(<LoginForm />);

    expect(container).toMatchSnapshot();
  });
});
