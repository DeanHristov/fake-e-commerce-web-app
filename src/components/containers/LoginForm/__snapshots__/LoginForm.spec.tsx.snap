// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container/Component <LoginForm {...} /> Should render the badge with a counter 1`] = `
<div>
  <div
    class="bg-white border border-gray-300 rounded-lg p-4 shadow-md inline-block relative undefined"
  >
    <h3
      class="text-base text-black font-bold pb-2"
    >
      Login
    </h3>
    <form>
      <div
        class="py-2 py-2 w-96"
      >
        <label
          class="block text-xs mb-1 cursor-pointer"
          for="email"
        >
          Enter an Email
        </label>
        <div
          class="relative w-full rounded-md h-12 border border-gray-500"
        >
          <input
            class="h-full px-6 w-full rounded-lg bg-white text-slate-700 ease-in duration-200 focus:bg-white"
            id="email"
            name="email"
            type="text"
            value="visitor@my-site.com"
          />
        </div>
      </div>
      <div
        class="py-2 py-2 w-96"
      >
        <label
          class="block text-xs mb-1 cursor-pointer"
          for="password"
        >
          Enter a password
        </label>
        <div
          class="relative w-full rounded-md h-12 border border-gray-500"
        >
          <input
            class="h-full px-6 w-full rounded-lg bg-white text-slate-700 ease-in duration-200 focus:bg-white"
            id="password"
            name="password"
            type="password"
            value="13579"
          />
        </div>
      </div>
      <button
        class="cursor-pointer bg-green-600 text-white font-bold uppercase bg-primary hover:bg-hover hover:shadow-lg transition transform active:scale-90 duration-200 ease-in w-full px-4 py-1.5 flex justify-center items-center gap-2 py-2 w-96 rounded-xl"
        type="submit"
      >
        Login
      </button>
    </form>
    <a
      class="text-blue-400 text-sm mt-4 text-center hover:underline"
      href="/reset-password"
    >
      Reset password
    </a>
  </div>
</div>
`;
